bundles/org.openhab.core.io.transport.upnp/src/main/java/org/openhab/core/io/transport/upnp/UpnpDeviceFinder.java

```diff
@@ -0,0 +1,170 @@
+import org.jupnp.model.types.UDN;
+import org.jupnp.registry.Registry;
+import org.jupnp.registry.RegistryListener;
+import org.openhab.core.common.ThreadPoolManager;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Reference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Component service that can be used to keep mis-behaving UPnP devices stored in the registry
+ * by sending a targeted M-SEARCH message for their UDN before their maxAge expires.
+ *
+ * @author Andrew Fiddian-Green - Initial Contribution
+ */
+@Component(immediate = true)
+public class UpnpDeviceFinder implements RegistryListener {
+
+    private static final String DEVICE_FINDER_THREADPOOL = "upnpDeviceFinder";
+    private static final int SEARCH_SCHEDULE_SECONDS = 60;
+
+    private final Logger logger = LoggerFactory.getLogger(UpnpDeviceFinder.class);
+    private final ScheduledExecutorService scheduler = ThreadPoolManager.getScheduledPool(DEVICE_FINDER_THREADPOOL);
+    private final Map<UDN, Future<?>> subscriptions = new ConcurrentHashMap<>();
+
+    private final UpnpService upnpService;
+
+    @Activate
+    public UpnpDeviceFinder(final @Reference UpnpService upnpService) {
+        this.upnpService = upnpService;
+    }
+
+    @Activate
+    public void activate() {
+        upnpService.getRegistry().addListener(this);
+        upnpService.getControlPoint().search();
+        upnpService.getControlPoint().search(new RootDeviceHeader());
+    }
+
+    @Deactivate
+    public void deactivate() {
+        upnpService.getRegistry().removeListener(this);
+        subscriptions.values().forEach(task -> task.cancel(false));
+        subscriptions.clear();
+    }
+
+    /**
+     * Cancel scheduled search (if any) for the given UDN.
+     */
+    private void cancelSearch(UDN udn) {
+        Future<?> task = subscriptions.get(udn);
+        if (task != null) {
+            task.cancel(false);
+        }
+    }
+
+    /**
+     * Schedule a search for the given device at a future time based on its maxAge.
+     */
+    private void scheduleSearch(RemoteDevice device) {
+        UDN udn = device.getIdentity().getUdn();
+        int delaySeconds = Math.max(SEARCH_SCHEDULE_SECONDS,
+                device.getIdentity().getMaxAgeSeconds() - SEARCH_SCHEDULE_SECONDS);
+        logger.debug("Scheduling search for {} in {} seconds", udn, delaySeconds);
+        cancelSearch(udn);
+        subscriptions.put(udn, scheduler.schedule(() -> {
+            logger.debug("Executing search for {}", udn);
+            upnpService.getControlPoint().search(new UDNHeader(udn));
+        }, delaySeconds, TimeUnit.SECONDS));
+    }
+
+    /**
+     * Add the given UPnP device UDN to the subscriptions list and execute an initial immediate search.
+     */
+    public void addUDN(UDN udn) {
```

Comment:
Hmm, i don't see where this is actually called?  right now i don't think any subscriptions are being added to that map.

Reply:
I see, this is meant to be called from external bindings?

Reply:
Yes. The addon has to load this component as a reference in its constructor. And then subscribe to it for every UDN that it wants to regularly search for.

Reply:
Sorry i missed that part, so ignore my previous test then.  I'll give this a try now as it was intended.

Reply:
This is running now, and i can see it scheduling checks, so 🤞

Reply:
So good news, it successfully registered the device before it expired. I'll keep on eye on it overnight .